### ISSUE TITLE
--randomize flag is more random than expected

### ISSUE THREAD

**Issue by davidsoncolin** [glinscott/leela-chess#185]
In leela-zero UCTSearch.cpp:296 we see:

```
    // Check whether to randomize the best move proportional
    // to the playout counts, early game only.
    auto movenum = int(m_rootstate.get_movenum());
    if (movenum < cfg_random_cnt) {
        m_root->randomize_first_proportionally();
    }
```

Compare this with the code in leela-chess UCTSearch.cpp:137

```
    // Check whether to randomize the best move proportional
    // to the playout counts.
    if (cfg_randomize) {
        m_root.randomize_first_proportionally();
    }
```

The result is that the --randomize flag applies the random selection heuristic to every move of every training game. The unfortunate result is that all of the training games are overly random. I would suggest that this flag should be set only for the opening moves?

**Comment by alreadydone** [comment]
AlphaGo Zero used randomness in the first 30 moves and AlphaZero used it all game long.

**Comment by davidsoncolin** [comment]
You're quite right. I hadn't picked up on that detail. [AlphaZero](https://arxiv.org/pdf/1712.01815.pdf) page 14:

> During training, [...] Moves are selected in proportion to the root visit count.

It makes perfect sense to train on that distribution, since we want it to learn the probability of selecting a move. Is anyone else astonished at the disparity between the training data and the evaluation though? I mean, in the training training data one side generally loses a couple of queens whereas in matchplay it's a 1600 elo...

I had imagined that the results of search during training would be a slightly improved version of the static evaluation, providing a direction for continual improvement. But in training the search seems considerably worse than the evaluation. So is what's really happening here that it is learning not to ever blunder queens ... by blundering a million queens?

It just seems to me that those gradients would be so much steeper if, for example, only one side was compromised.

**Comment by Dorus** [comment]
It is very interesting that these move random moves actually help the network train, This seems counter intuitive, but remember that 1) The game result will be right most of the time. This is enough for the value net to learn. 2) The policy net will still train on the move that it found, not the move that it played. So the policy net is still trained optimal. 3) it plays more random, but only moves that the policy net suggested, eventually it will always play the right move as the policy net will no longer suggest bad moves. At that point the value net will be trained even better. 

I realize there is a also noise added to the search so the policy net can never reject all bad moves (since noise add the bad moves back in), but this is required for training to be able to find new good moves too, and we can refer back to (1) that the majority of the games will still be correct anyway.

**Comment by killerducky** [comment]
This is being done as specified in the AZ paper.
